test/nlp_expr.jl

```diff
@@ -0,0 +1,348 @@
+    return
+end
+
+function test_nonlinear_matrix_algebra()
+    model = Model()
+    @variable(model, X[1:3, 1:3], Symmetric)
+    @objective(model, Max, sum(X^4 .- X^3))
+    nlp = nonlinear_model(model)
+    Y = [0.1 0.2 0.3; 0.2 0.4 0.5; 0.3 0.5 0.6]
+    data = Dict(index(X[i, j]) => Y[i, j] for i in 1:3 for j in 1:i)
+    obj = MOI.Nonlinear.evaluate(data, nlp, nlp.objective)
+    @test obj ≈ sum(Y^4 .- Y^3)
+    return
+end
+
+"""
+This test checks that we can work with expressions of arbitrary depth. Don't use
+recursion!
+"""
+function test_recursion_stackoverflow()
+    model = Model()
+    @variable(model, x)
+    expr = sin(x)
+    for _ in 1:20_000
+        expr = sin(expr)
+    end
+    @test @objective(model, Min, expr) isa NonlinearExpr
+    @test string(expr) isa String
+    return
+end
+
+function test_nlparameter_interaction()
+    model = Model()
+    @variable(model, x)
+    @NLparameter(model, p == 1)
+    e = x + p
+    @test e isa NonlinearExpr
+    @test string(e) == "+(x, $p)"
+    return
+end
+
+function test_nlexpression_interaction()
+    model = Model()
+    @variable(model, x)
+    @NLexpression(model, expr, sin(x))
+    e = x + expr
+    @test e isa NonlinearExpr
+    @test string(e) == "+(x, $expr)"
+    return
+end
+
+function test_nlobjective_with_nlexpr()
+    model = Model()
+    @variable(model, x)
+    y = sin(x)
+    @NLobjective(model, Min, y^2)
+    nlp = nonlinear_model(model)
+    @test isequal_canonical(jump_function(model, nlp.objective), sin(x)^2)
+    return
+end
+
+function test_nlconstraint_with_nlexpr()
+    model = Model()
+    @variable(model, x)
+    y = sin(x)
+    @NLconstraint(model, c, y^2 <= 1)
+    nlp = nonlinear_model(model)
+    @test isequal_canonical(
+        jump_function(model, nlp.constraints[index(c)].expression),
+        sin(x)^2 - 1,
+    )
+    return
+end
+
+function test_jump_function_nonlinearexpr()
+    model = Model()
+    @variable(model, x)
+    @NLparameter(model, p == 1)
+    @NLexpression(model, expr1, sin(p + x))
+    @NLexpression(model, expr2, sin(expr1))
+    nlp = nonlinear_model(model)
+    @test string(jump_function(model, nlp[index(expr1)])) == "sin(+($p, $x))"
+    @test string(jump_function(model, nlp[index(expr2)])) == "sin($expr1)"
+    return
+end
+
+function test_constraint_object()
+    model = Model()
+    @variable(model, x)
+    y = sin(x)
+    @NLconstraint(model, c, y^2 <= 1)
+    con = constraint_object(c)
+    @test isequal_canonical(con.func, sin(x)^2 - 1.0)
+    @test con.set == MOI.LessThan(0.0)
+    return
+end
+
+function test_expr_mle()
+    data = [1.0, 2.0, 4.0, 8.0]
+    n = length(data)
+    model = Model()
+    @variable(model, x)
+    @variable(model, y)
+    obj = @expression(
+        model,
+        n / 2 * log(1 / (2 * y^2)) -
+        sum((data[i] - x)^2 for i in 1:n) / (2 * y^2)
+    )
+    @test string(obj) ==
+          "-(*(2.0, log(/(1.0, 2 $(y^2)))), /(4 $(x^2) - 30 x + 85, 2 $(y^2)))"
```

Comment:
So after a looong, slightly unexpected detour through MutableArithmetics, we now have this working nicely.